ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java

```diff
@@ -167,6 +174,52 @@ public void execute(Input input, ActionListener<Output> listener) {
 
     }
 
+    /**
+     * save root interaction and start execute the agent
+     * @param listener callback listener
+     * @param memory memory instance
+     * @param inputDataSet input
+     * @param mlAgent agent to run
+     */
+    private void saveRootInteractionAndExecute(
+        ActionListener<Object> listener,
+        ConversationIndexMemory memory,
+        RemoteInferenceInputDataSet inputDataSet,
+        MLAgent mlAgent
+    ) {
+        String appType = mlAgent.getAppType();
+        String question = inputDataSet.getParameters().get(QUESTION);
+        String regenerateInteractionId = inputDataSet.getParameters().get(REGENERATE_INTERACTION_ID);
+        // Create root interaction ID
+        ConversationIndexMessage msg = ConversationIndexMessage
+            .conversationIndexMessageBuilder()
+            .type(appType)
+            .question(question)
+            .response("")
+            .finalAnswer(true)
+            .sessionId(memory.getConversationId())
+            .build();
+        memory.save(msg, null, null, null, ActionListener.<CreateInteractionResponse>wrap(interaction -> {
+            log.info("Created parent interaction ID: " + interaction.getId());
+            inputDataSet.getParameters().put(PARENT_INTERACTION_ID, interaction.getId());
+            // only delete previous interaction when new interaction created
+            ActionListener<Object> finalListener = regenerateInteractionId == null
+                ? listener
+                : ActionListener
+                    .runBefore(
+                        listener,
+                        () -> memory.getMemoryManager().deleteInteraction(regenerateInteractionId, ActionListener.wrap(deleted -> {}, e -> {
```

Comment:
What about the children interactions? Do we delete them?

Reply:
Yes, both child interaction and parent interaction are deleted.

Reply:
renamed to `deleteInteractionAndTrace`